File: Data Structures & Algorithms/same-binary-tree/submission-10.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        //both null, vacously true
        if(p == null && q == null){
            return true;
        }

        //if only 1 is null or not same value --> false
        if(p == null || q == null || p.val != q.val){
            return false;
        }

        //otherwise, check the subtrees
        return isSameTree(p.left,q.left) && isSameTree(p.right, q.right);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-10.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        def dfs(node):
            if not node:
               return (0,True)
            
            leftDepth, isLeftBal = dfs(node.left)
            rightDepth, isRightBal = dfs(node.right)
            depth = 1 + max(leftDepth, rightDepth)

            if not isLeftBal or not isRightBal:
               return (depth, False)
            if abs(leftDepth - rightDepth) > 1:
               return (depth, False)
            
            return (depth, True)

        return dfs(root)[1]    
        